### PR TITLE
Separate ingress and egress default deny modes with better control

### PR DESCRIPTION
**Pull request by manalibhutiyani**
Proposal: When to put ingress and egress of an endpoint into default deny

Fixes Issue : #2002

The current behaviour puts an endpoint into default deny at ingress
whenever there is a policy rule that selects the endpoint.
This behaviour can be modified by running the agent in "always"
policy enforcement mode in which case all endpoints are always in default deny mode.

With the introduction of egress rules, it is unclear whether
a policy rule should put an endpoint into default deny at ingress, egress, or both.

Proposal:

Basic rules:

1. If any rule selects an endpoint and the rule has an ingress section,
   the endpoint goes into default deny at ingress
2. If any rule selects an endpoint and the rule has an egress section,
   the endpoint goes into default deny at egress.
3. An empty endpoint label selector selects all endpoints
4. An empty CIDR block allows nothing

Tasks:
1. Add endpoint fields to store default deny for ingress
   and egress (replace current logic which is policy on/off by endpoint)
2. Add fields to display default deny on ingress and egress
3. Update cilium endpoint list to show ingress and egress instead of just "Policy enabled"
4. Verify datapath and enforce correctly

Signed-off-by: Manali Bhutiyani manali@covalent.io

```release-note
Separate ingress and egress default deny modes with better control
```

### CODE REVIEW

File: cilium/cmd/endpoint_list.go [hunk 1]
```diff
@@ -45,14 +45,24 @@ func init() {
...
-	var isPolicyEnabled string
-	if swag.BoolValue(ep.PolicyEnabled) {
-		isPolicyEnabled = "Enabled"
-	} else {
-		isPolicyEnabled = "Disabled"
+	var isIngressPolicyEnabled string
+	var isEgressPolicyEnabled string
+	switch swag.StringValue(ep.PolicyEnabled) {
+	case "none":
+		isIngressPolicyEnabled = "Disabled"
```
Comment (ianvernon) [inline cilium/cmd/endpoint_list.go]: Nit - make "Disabled" and "Enabled" constants

File: pkg/policy/repository.go [hunk 2]
```diff
@@ -366,19 +366,34 @@ func (p *Repository) GetJSON() string {
 // rule with labels matching the labels in the provided LabelArray.
 //
 // Must be called with p.Mutex held
-func (p *Repository) GetRulesMatching(labels labels.LabelArray, includeEntities bool) bool {
-
+func (p *Repository) GetRulesMatching(labels labels.LabelArray, includeEntities bool) (ingressMatch bool, egressMatch bool) {
```
Comment (ianvernon) [inline pkg/policy/repository.go:376]: update comment specifying what includeEntities means

**Review by ianvernon — CHANGES_REQUESTED** [review]
First pass looks good for the most part. This might break some of our tests because unfortunately they use `awk` and `grep` to check for whether policy enforcement is enabled or disabled, so make sure to update those as part of this patch.

File: pkg/policy/repository.go (hunk 2 above)
Comment (manalibhutiyani) [reply pkg/policy/repository.go:376]: done

**Comment by manalibhutiyani** [comment]
@ianvernon Updated all the CI tests to account for the additional output in `cilium endpoint list`

File: pkg/endpoint/bpf.go [hunk 3]
```diff
@@ -180,7 +180,9 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	fw.WriteString(" */\n\n")
 
 	// If there hasn't been a policy calculated yet, we need to be sure we drop all packets, but only if policy enforcement is enabled for the endpoint / daemon.
-	if !e.PolicyCalculated && e.Opts.IsEnabled(OptionPolicy) && owner.PolicyEnforcement() != NeverEnforce {
+	if !e.PolicyCalculated &&
+		(e.Opts.IsEnabled(OptionIngressPolicy) || e.Opts.IsEnabled(OptionEgressPolicy)) &&
```
Comment (tgraf) [inline pkg/endpoint/bpf.go]: While reviewing this, I think the existing logic was actually wrong. We want to go into `DROP_ALL` on !e.PolicyCalculated && owner.PolicyEnforcement() != NeverEnforce`, i.e. when there is a chance that policy may be enabled for this endpoint but we have not calculated it yet. If we are in auto mode, there is no chance for either ingress or egress to be enabled before the calculation has completed for the first time.

File: api/v1/openapi.yaml [hunk 4]
```diff
@@ -794,8 +794,8 @@ definitions:
         description: Policy information of endpoint
         "$ref": "#/definitions/EndpointPolicy"
       policy-enabled:
-        description: Whether policy enforcement is enabled or not
-        type: boolean
+        description: Whether policy enforcement is enabled (ingress, egress, both or none)
+        type: string
```
Comment (tgraf) [inline api/v1/openapi.yaml:798]: Add the following here:
```
enum:
- ingress
- egress
- both
- none
```

This will also define golang constants that you can then use.

File: cilium/cmd/endpoint_list.go [hunk 5]
```diff
@@ -45,14 +45,24 @@ func init() {
...
+		isEgressPolicyEnabled = "Disabled"
+	case "both":
```
Comment (tgraf) [inline cilium/cmd/endpoint_list.go]: These should be using the constants generated from the before mentioned enum

File: pkg/endpoint/endpoint.go [hunk 6]
```diff
@@ -412,7 +419,16 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
 		currentState = models.EndpointStateNotReady
 	}
 
-	policyEnabled := e.Opts.IsEnabled(OptionPolicy)
+	policyIngressEnabled := e.Opts.IsEnabled(OptionIngressPolicy)
+	policyEgressEnabled := e.Opts.IsEnabled(OptionEgressPolicy)
+
+	if policyIngressEnabled && policyEgressEnabled {
+		policy = "both"
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go]: These should be using the constants generated from the before mentioned enum

**Review by tgraf — CHANGES_REQUESTED** [review]

File: api/v1/openapi.yaml (hunk 4 above)
Comment (manalibhutiyani) [reply api/v1/openapi.yaml:798]: done

**Review by manalibhutiyani — COMMENTED** [review]
@ianvernon and @tgraf :Addressed all the review comments. Please take a look.

File: pkg/policy/repository.go [hunk 7]
```diff
@@ -369,6 +369,8 @@ func (p *Repository) GetJSON() string {
 
 // GetRulesMatching returns whether any of the rules in a repository contain a
 // rule with labels matching the labels in the provided LabelArray.
+// If includeEntities is true, we check if repository contains rules matching
```
Comment (aanm) [inline pkg/policy/repository.go:372]: The commit message seems wrong

File: pkg/endpoint/policy.go [hunk 8]
```diff
@@ -554,24 +554,27 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
 
 	ingress, egress := owner.EnableEndpointPolicyEnforcement(e)
 
-	if ingress {
-		opts[OptionIngressPolicy] = "disabled"
-	}
+	opts[OptionIngressPolicy] = "disabled"
+	opts[OptionEgressPolicy] = "disabled"
```
Comment (aanm) [inline pkg/endpoint/policy.go]: Small nit to use consts for `disabled` and `enabled`?

File: pkg/policy/repository.go (hunk 7 above)
Comment (manalibhutiyani) [reply pkg/policy/repository.go:372]: Changed the wording a bit to make it more clear.

**Comment by manalibhutiyani** [comment]
@aanm : Addresses your review comments.

File: pkg/endpoint/bpf.go (hunk 3 above)
Comment (manalibhutiyani) [reply pkg/endpoint/bpf.go]: Hi Thomas changing this to 
 ```
 if !e.PolicyCalculated && owner.PolicyEnforcement() != NeverEnforce {
            fw.WriteString("#define DROP_ALL\n")
     } 
```
as you suggested, gives a cilium - consul timeout (>30s timeout error). I have messaged you offline with the logs. I am reverting this back to original logic and debugging it right now. Any suggestions why this could be the case? I am not sure if that's the reason why the existing logic was that way.

Comment (manalibhutiyani) [reply pkg/endpoint/bpf.go]: @tgraf  I reconfirmed by running tests locally too, that this check is indeed causing the timeout. I will continue debugging it, but if this feature is blocking the release, I can follow this up in a separate PR/GH issue. Please let me know!

File: cilium/cmd/endpoint_list.go [hunk 9]
```diff
@@ -66,20 +82,21 @@ func listEndpoints() {
...
+		enforcementTitle   = "ENFORCEMENT"
 	)
 
 	if !noHeaders {
-		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t\n",
-			endpointTitle, policyTitle, labelsIDTitle, labelsDesTitle, ipv6Title, ipv4Title, statusTitle)
-		fmt.Fprintf(w, "\t%s\t\t\t\t\t\t\n", enforcementTitle)
+		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t\n",
+			endpointTitle, policyIngressTitle, policyEgressTitle, labelsIDTitle, labelsDesTitle, ipv6Title, ipv4Title, statusTitle)
+		fmt.Fprintf(w, "\t%s\t%s\t\t\t\t\t\n", enforcementTitle, enforcementTitle)
```
Comment (rlenglet) [inline cilium/cmd/endpoint_list.go:99]: Why the change to this line? I think it should be reverted.

**Review by rlenglet — APPROVED** [review]

**Comment by tgraf** [comment]
Rebased

**Review by tgraf — APPROVED** [review]

File: cilium/cmd/endpoint_list.go (hunk 9 above)
Comment (rlenglet) [reply cilium/cmd/endpoint_list.go:99]: Nevermind.

**Review by rlenglet — APPROVED** [review]

File: api/v1/openapi.yaml (hunk 4 above)
Comment (joestringer) [reply api/v1/openapi.yaml:798]: Just FYI this type of change breaks the API, so if you use `cilium` client from before this change to perform `cilium endpoint get` with a daemon that has this commit, then the `cilium endpoint get` returns:

```
Error: Cannot get endpoint 62006: json: cannot unmarshal bool into Go struct field Endpoint.policy-enabled of type string
```

Comment (manalibhutiyani) [reply api/v1/openapi.yaml:798]: @joestringer I can raise a separate issue for it and fix it.

Comment (joestringer) [reply api/v1/openapi.yaml:798]: @tgraf @manalibhutiyani I don't know if we're that concerned right now, mainly I just wanted to highlight that this kind of change will not be OK post-1.0.

Comment (manalibhutiyani) [reply api/v1/openapi.yaml:798]: I agree. But this change was required to make cilium aligned to K8s behavior. So making this api change was un-avoidable. But I agree, with v1.0 we need to be more mindful before breaking api contracts
